google-cloud-firestore/lib/google/cloud/firestore/query.rb

```diff
@@ -0,0 +1,617 @@
+# Copyright 2017 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+require "google/cloud/firestore/v1beta1"
+require "google/cloud/firestore/document_snapshot"
+require "google/cloud/firestore/convert"
+
+module Google
+  module Cloud
+    module Firestore
+      ##
+      # # Query
+      #
+      # Represents a query to the Firestore API.
+      #
+      # Instances of this class are immutable. All methods that refine the query
+      # return new instances.
+      #
+      # @example
+      #   require "google/cloud/firestore"
+      #
+      #   firestore = Google::Cloud::Firestore.new
+      #
+      #   # Create a query
+      #   query = firestore.col(:cities).select(:population)
+      #
+      #   query.get do |city|
+      #     puts "#{city.document_id} has #{city[:population]} residents."
+      #   end
+      #
+      class Query
+        ##
+        # @private The parent path for the query.
+        attr_accessor :parent_path
+
+        ##
+        # @private The Google::Firestore::V1beta1::Query object.
+        attr_accessor :query
+
+        ##
+        # @private The firestore client object.
+        attr_accessor :client
+
+        ##
+        # Restricts documents matching the query to return only data for the
+        # provided fields.
+        #
+        # @param [String, Symbol, Array<String|Symbol>] fields A field mask to
+        #   filter results with and return only the specified fields. One or
+        #   more field paths can be specified.
+        #
+        # @return [Query] New query with `select` called on it.
+        #
+        # @example
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   # Get a collection reference
+        #   cities_col = firestore.col "cities"
+        #
+        #   # Create a query
+        #   query = cities_col.select(:population)
+        #
+        #   query.get do |city|
+        #     puts "#{city.document_id} has #{city[:population]} residents."
+        #   end
+        #
+        def select *fields
```

Comment:
@schmidt-sebastian I have a question about this. Can you select a nested field path? If so, are the rules for formatting nested field paths the same rules for formatting that `update` used? e.g.

```
foo.`bar.baz`.biz
```

Reply:
Yes, the query API accepts this field path format for where, orderBy and select.

Reply:
This has been implemented and the documentation made more explicit.